src/registration.rs

```diff
@@ -0,0 +1,192 @@
+    #[serde(default)]
+    pub admin: bool,
+
+    pub localname: String,
+
+    // if the password isn't provided, use the localname
+    #[serde(default)]
+    password: Option<String>,
+}
+
+impl User {
+    pub fn password(&self) -> String {
+        match &self.password {
+            Some(password) => password.clone(),
+            None => self.localname.clone(),
+        }
+    }
+}
+
+/// Register a user using the admin api and a registration shared secret.
+/// The base_url is the Scheme and Authority of the URL to access synapse via.
+/// Returns a RegistrationResponse if registration succeeded, otherwise returns an error.
+pub async fn register_user(
+    base_url: &str,
+    registration_shared_secret: &str,
+    user: &User,
+) -> Result<RegistrationResponse, Error> {
+    #[derive(Debug, Deserialize)]
+    struct GetRegisterResponse {
+        nonce: String,
+    }
+    let registration_url = format!("{}/_synapse/admin/v1/register", base_url);
+    debug!(
+        "Registration shared secret: {}, url: {}, user: {:#?}",
+        registration_shared_secret, registration_url, user
+    );
+    let nonce = reqwest::get(&registration_url)
+        .await?
+        .json::<GetRegisterResponse>()
+        .await?
+        .nonce;
+    let mut mac =
+        HmacSha1::new_from_slice(registration_shared_secret.as_bytes()).map_err(|err| {
+            anyhow!(
+                "Couldn't use the provided registration shared secret to create a hmac: {}",
+                err
+            )
+        })?;
+    mac.update(
+        format!(
+            "{nonce}\0{username}\0{password}\0{admin}",
+            nonce = nonce,
+            username = user.localname,
+            password = user.password(),
+            admin = if user.admin { "admin" } else { "notadmin" }
+        )
+        .as_bytes(),
+    );
+    let registration_payload = RegistrationPayload {
+        nonce,
+        username: user.localname.to_string(),
+        displayname: user.localname.to_string(),
+        password: user.password(),
+        admin: user.admin,
+        mac: HEXLOWER.encode(&mac.finalize().into_bytes()),
+    };
+    debug!(
+        "Sending payload {:#?}",
+        serde_json::to_string_pretty(&registration_payload)
+    );
+    let client = reqwest::Client::new();
+    let response = client
+        .post(&registration_url)
+        .json(&registration_payload)
+        .send()
+        .await?;
+    match response.status() {
+        StatusCode::OK => {
+            let registration_info = response.json::<RegistrationResponse>().await?;
+            Ok(registration_info)
+        }
+        _ => {
+            let body = response.json::<ErrorResponse>().await?;
+            Err(anyhow!(
+                "Homeserver responded with errcode: {}, error: {}",
+                body.errcode,
+                body.error
+            ))
+        }
+    }
+}
+
+pub async fn login(base_url: &str, user: &User) -> Result<RegistrationResponse, Error> {
```

Comment:
Do we really need the full `RegistrationResponse`? A token would be sufficient, no?

Reply:
we don't use the access tokens either, so it is a little bit weird to do either